torch/nested/_internal/sdpa.py

```diff
@@ -0,0 +1,723 @@
+        if backend == SDPBackend.EFFICIENT_ATTENTION:
+            if math_sdp_enabled() and _can_use_math_sdpa_jagged(params):
+                return SDPBackend.MATH
+
+    log.warning("Memory efficient kernel not used because:")
+    can_use_efficient_attention(params, debug=True)
+    _can_use_efficient_sdpa_jagged(params, debug=True)
+    log.warning("Flash attention kernel not used because:")
+    can_use_flash_attention(params, debug=True)
+    _can_use_flash_sdpa_jagged(params, debug=True)
+    raise ValueError("No available kernel. Aborting execution.")
+
+
+def _cumulative_and_max_seq_len_nnz(qkv):
+    """
+    This function is used to calculate two pieces of metadata that are needed
+    for use with flash-attention and efficient_attention kernels. They are the
+    cumulative sequence_length over a batch of sequences and the maximum
+    sequence length.
+
+    @return A tuple of cumulative sequence lengths and the maximum sequence
+    length, and the last element in the cumulative_sequence_lengths
+    """
+    if not qkv.is_nested:
+        raise ValueError("QKV must be nested for flash cumulative_seq_len calculation.")
+
+    if qkv.lengths() is None:
+        cumulative_seqlen = qkv.offsets().to(dtype=torch.int32, device=qkv.device)
+        max_seqlen = torch.max(qkv.offsets().diff()).item()
+    else:
+        cumulative_seqlen = (
+            qkv.lengths().cumsum(0).to(dtype=torch.int32, device=qkv.device)
+        )
+        batch_size = qkv.size(0)
+        max_seqlen = qkv.values().size(0) / batch_size
+    return cumulative_seqlen, max_seqlen, cumulative_seqlen[-1]
+
+
+def _is_safe_to_get_storage_as_tensor(tensor):
+    """
+    This function checks if a nested tensor is valid for
+    use with the flash-attention and efficient_attention kernels without
+    needing to call contiguous on the nested tensor input.
+    It checks that the storage offsets' adjacent_differences are a constant
+    mutiple of the previous tensor in the nested tensor and that the strides
+    are monitonically decreasing. This check is done after calling transpose on
+    the nested tensor. Resulting in a Nt of shape [bsz, {seq_len}, num_heads, dim]
+
+    @return A boolean indicating if contiguous needs to be called for input
+    """
+    offsets = tensor.offsets()
+    strides = tensor._stride
+
+    n_tensors = offsets.size(0) - 1
+    if n_tensors <= 1:
+        return True
+
+    # Check initially that the tensor strides are in strictly descending order
+    prev_stride = strides[1]
+    for stride in strides[2:]:
+        if prev_stride <= stride:
+            # This would mean that the last stride is greater than the seq_len
+            # stride
+            return False
+        prev_stride = stride
+
+    # Check the offsets are a constant multiple from the previous numels
+    offset_constant = offsets[1] - offsets[0]
+    for i in range(2, len(offsets)):
+        current_offset_constant = offsets[i] - offsets[i - 1]
```

Comment:
talked offline if we only support the vanilla dense jagged layout then this might be simpler

Reply:
yup, looking at this again

Reply:
ok, removed the check, as it wasn't doing what it was supposed to anyways